File: clrs/algorithms/greedy_elements.py
class Item:
  def __init__(self, value, weight):
    self.value = value
    self.weight = weight

def fractional_knapsack(capacity, items):
  items.sort(key=lambda x: x.value / x.weight, reverse = True)

  total_value = 0.0

  for item in items:
    if capacity >= item.weight:
      capacity -= item.weight
      total_value += item.value
    else:
      fraction = capacity / item.weight
      total_value += item.value * fraction
      break
  
  return total_value

def knapsack_0_1(v, w, n, W):
  c = [[0] * (W + 1) for _ in range(n + 1)]

  for i in range(1, n + 1):
    for capacity in range(1, W + 1):
      if w[i - 1] <= capacity:
        c[i][capacity] = max(
          v[i - 1] + c[i - 1][capacity - w[i - 1]], 
          c[i - 1][capacity]
        )
      else:
        c[i][capacity] = c[i - 1][capacity]

  return c[n][W]

if __name__ == '__main__':
  max_capacity = 50

  items_list = [Item(60, 10), Item(100, 20), Item(120, 30)]
  values = [item.value for item in items_list]
  weights = [item.weight for item in items_list]

  max_val = fractional_knapsack(max_capacity, items_list)
  max_0_1_val = knapsack_0_1(values, weights, len(items_list), max_capacity)

  print(f"Maximum value in fractional Knapsack: {max_val:.2f}")
  print(f"Maximum value in 0/1 Knapsack: {max_0_1_val}")

File: clrs/algorithms/greedy_elements.js
class Item {
  constructor(value, weight) {
    this.value = value;
    this.weight = weight;
  }
}

function fractionalKnapsack(capacity, items) {
  items.sort((a, b) => b.value / b.weight - a.value / a.weight);
  let totalValue = 0;

  for (const item of items) {
    if (capacity >= item.weight) {
      capacity -= item.weight;
      totalValue += item.value;
    } else {
      const fraction = capacity / item.weight;
      totalValue += item.value * fraction;
      break;
    }
  }

  return totalValue;
}

function knapsack01(values, weights, n, maxCapacity) {
  const best = Array.from({ length: n + 1 }, () =>
    Array(maxCapacity + 1).fill(0),
  );

  for (let i = 1; i <= n; i += 1) {
    for (let capacity = 1; capacity <= maxCapacity; capacity += 1) {
      if (weights[i - 1] <= capacity) {
        best[i][capacity] = Math.max(
          values[i - 1] + best[i - 1][capacity - weights[i - 1]],
          best[i - 1][capacity],
        );
      } else {
        best[i][capacity] = best[i - 1][capacity];
      }
    }
  }

  return best[n][maxCapacity];
}

const maxCapacity = 50;
const items = [new Item(60, 10), new Item(100, 20), new Item(120, 30)];
const values = items.map((item) => item.value);
const weights = items.map((item) => item.weight);

const fractionalMax = fractionalKnapsack(maxCapacity, items);
const zeroOneMax = knapsack01(values, weights, items.length, maxCapacity);

console.log(
  `Maximum value in fractional Knapsack: ${fractionalMax.toFixed(2)}`,
);
console.log(`Maximum value in 0/1 Knapsack: ${zeroOneMax}`);
